File: docs/Sword2offer/CH4-解题思路/ES6/bfs.js
const rPreOrder = Symbol('preorder')

class Node {
  constructor (value) {
    this.value = value
    this.left = null
    this.right = null
  }
}

class Tree {
  constructor () {
    this.head = null
  }
  get getHead () {
    return this.head
  }
  setHead (node) {
    this.head = node
  }
  [rPreOrder] (node, callback) {
    callback(node.value)
    if (node.left) {
      this[rPreOrder](node.left, callback)
    }
    if (node.right) {
      this[rPreOrder](node.right, callback)
    }
  }
  preOrder (callback) {
    if (!this.head) {
      return null
    }
    this[rPreOrder](this.head, callback)
  }
  bfsOrder (callback) {
    let queue = [this.head]
    while (queue.length) {
      let node = queue.shift()
      callback(node.value)
      if (node.left) {
        queue.push(node.left)
      }
      if (node.right) {
        queue.push(node.right)
      }
    }
  }
}

let node1 = new Node(8)
let node2 = new Node(6)
let node3 = new Node(10)
let node4 = new Node(5)
let node5 = new Node(7)
let node6 = new Node(9)
let node7 = new Node(11)

let treeA = new Tree()
node1.left = node2
node1.right = node3
node2.left = node4
node2.right = node5
node3.left = node6
node3.right = node7
treeA.setHead(node1)
// 测试是否构建正确
treeA.bfsOrder(function (value) {
  console.log(value)
})

File: docs/Sword2offer/CH4-解题思路/Python/bfs.py
class Node:
  def __init__(self, value):
    self.value = value
    self.left = None
    self.right = None


class Tree:
  def __init__(self):
    self.head = None

  def __recurvise_preorder(self, node):
    print(node.value)
    if node.left:
      self.__recurvise_preorder(node.left)
    if node.right:
      self.__recurvise_preorder(node.right)

  def preorder(self):
    if not self.head:
      return
    self.__recurvise_preorder(self.head)

  def bfsOrder(self, callback):
    queue = [self.head]
    while len(queue):
      current = queue[0]
      del queue[0]
      callback(current.value)
      if current.left:
        queue.append(current.left)
      if current.right:
        queue.append(current.right)


node1 = Node(8)
node2 = Node(6)
node3 = Node(10)
node4 = Node(5)
node5 = Node(7)
node6 = Node(9)
node7 = Node(11)

node1.left = node2
node1.right = node3
node2.left = node4
node2.right = node5
node3.left = node6
node3.right = node7

treeA = Tree()
treeA.head = node1

if __name__ == '__main__':
  def show(value):
    print(value)
  treeA.bfsOrder(show)
